pallets/author-filter/src/lib.rs

```diff
@@ -0,0 +1,151 @@
+//! currently staked accounts from pallet stake, and filters them down to a pseudorandom subset.
+//! The current technique gives no preference to any particular author. In the future, we could
+//! disfavor authors who are authoring a disproportionate amount of the time in an attempt to
+//! "even the playing field".
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::pallet;
+
+pub use pallet::*;
+
+#[pallet]
+pub mod pallet {
+
+	use frame_support::pallet_prelude::*;
+	use frame_support::traits::Randomness;
+	use frame_support::traits::Vec;
+	use frame_system::pallet_prelude::*;
+	use sp_core::H256;
+	use sp_runtime::Percent;
+
+	/// The Author Filter pallet
+	#[pallet::pallet]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	/// Configuration trait of this pallet.
+	#[pallet::config]
+	pub trait Config:
+		frame_system::Config + stake::Config + cumulus_parachain_upgrade::Config
+	{
+		/// The overarching event type
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		/// Deterministic on-chain pseudo-randomness used to do the filtering
+		type RandomnessSource: Randomness<H256>;
+	}
+
+	// This code will be called by the author-inherent pallet to check whether the reported author
+	// of this block is eligible at this height. We calculate that result on demand and do not
+	// record it instorage (although we do emit a debugging event for now).
+	// This implementation relies on the relay parent's block number from the validation data
+	// inherent. Therefore the validation data inherent **must** be included before this check is
+	// performed. Concretely the validataion data inherent must be included before the author
+	// inherent.
+	impl<T: Config> author_inherent::CanAuthor<T::AccountId> for Pallet<T> {
+		fn can_author(account: &T::AccountId) -> bool {
+			let mut staked: Vec<T::AccountId> = stake::Module::<T>::validators();
+
+			let num_eligible = EligibleRatio::<T>::get().mul_ceil(staked.len());
+			let mut eligible = Vec::with_capacity(num_eligible);
+
+			// Grab the relay parent height as a temporary source of relay-based entropy
+			let validation_data = cumulus_parachain_upgrade::Module::<T>::validation_data()
+				.expect("validation data was set in parachain system inherent");
+			let relay_height = validation_data.persisted.block_number;
+
+			for i in 0..num_eligible {
+				// A context identifier for grabbing the randomness. Consists of three parts
+				// - The constant string *b"filter" - to identify this pallet
+				// - The index `i` when we're selecting the ith eligible author
+				// - The relay parent block number so that the eligible authors at the next height
+				//   change. Avoids liveness attacks from colluding minorities of active authors.
+				// Third one will not be necessary once we dleverage the relay chain's randomness.
+				let subject: [u8; 8] = [
+					b'f',
+					b'i',
+					b'l',
+					b't',
+					b'e',
+					b'r',
+					i as u8,
+					relay_height as u8,
```

Comment:
when the relay block number > u8::max(), will it just always put u8::max() here?

Reply:
Good question. I confess I did not know the answer to this before I explored it. But this playgroud demonstrates that when you do aprimitive case (with `as`) from `u32` to `u8` it just keeps the least significant 8 bits. Which means the u8 value wraps around. Which is what we want here.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7b61c262fc21e2d2915d942f6b4363ae